File: Python/complex root finder/find_number_of_roots.py
# -*- coding: utf-8 -*-
"""
Created on Sat Oct  5 20:47:51 2024

@author: RensvanLeijden
"""

# %%

import numpy as np
from scipy.integrate import simps
from scipy.fft import fft, ifft

# %%

def find_number_of_roots(f, Z, option_der='numerical', df=None, **kwargs):
    """
    Estimates the number of roots of a complex-valued function `f` in a given domain.
    
    Parameters:
    f : callable
        The complex-valued function for which roots need to be found.
    Z : numpy.ndarray
        A path along which the function is evaluated.
    option_der : str, optional
        Method to use for calculating the derivative. Options are 'symbolic', 'numerical', 'argp', or 'rootsofunity'.
    df : callable, optional
        Derivative function of `f`, used if `option_der` is 'symbolic'.
    kwargs : dict, optional
        Additional arguments required for specific options like 'argp' or 'rootsofunity'.
    
    Returns:
    Nroots : int
        The estimated number of roots.
    """

    # Calculate the number of roots based on the chosen derivative method
    if option_der == 'symbolic' and df is not None:
        # Use the provided symbolic derivative
        Nroots = simps(df(Z) / f(Z), Z) / (2 * np.pi * 1j)
    
    elif option_der == 'numerical':
        # Numerically calculate the derivative
        dfdz = np.diff(f(Z)) / np.diff(Z)
        Nroots = simps(dfdz / f(Z[:-1]), Z[:-1]) / (2 * np.pi * 1j)
    
    elif option_der == 'argp':
        # Derivation without explicit derivative (Peter & Kravanja method)
        Npoints = kwargs.get('Npoints', len(Z))
        Args = np.zeros(Npoints, dtype=np.complex_)
        for ii in range(Npoints - 1):
            Args[ii] = np.angle(f(Z[ii + 1]) / f(Z[ii]))
        Args[-1] = np.angle(f(Z[0]) / f(Z[-1]))
        Nroots = np.sum(Args) / (2 * np.pi)
    
    elif option_der == 'rootsofunity':
        z0 = kwargs['z0']
        r0 = kwargs['r0']
        step_size = 1e-6
        Npoints = int(np.round(2 * np.pi * r0 / step_size))
        Npoints = min(Npoints, 10000)  # Keep within max limit
        k = np.arange(Npoints)
        Z = z0 + r0 * np.exp(2 * np.pi * 1j * k / Npoints)
        
        n = Npoints
        fk = f(Z)
        c = fft(fk) / n
        cp = np.arange(1, n) * c[1:]
        ppzk = n * ifft(np.concatenate(([0], cp))) / r0

        # By Austin / Kravanja method
        Nroots = np.real(np.mean(Z * ppzk / fk))
    
    else:
        raise ValueError(f"Unknown option_der: {option_der}")

    # Handle branch cuts and adjust Nroots accordingly
    if np.isnan(Nroots) or np.isinf(Nroots):
        return 0

    # Check for imaginary value, indicating crossing a branch cut
    if round(np.imag(Nroots) / (1 / np.pi)) != 0:
        N_BC = round(np.imag(Nroots) / (1 / np.pi))
        
        if round(np.real(Nroots)) == 0:
            return 0
        
        if np.mod(np.real(Nroots), 1) != 0:
            # When values are 1.5, 2.5, etc.
            Nroots = round(np.real(Nroots), 2) + 1j * N_BC
        else:
            # When value is 0.5 (indicating one branch point)
            Nroots = round(np.real(Nroots), 1) + 1j * N_BC
        
        return Nroots

    # Round Nroots to the nearest integer
    return int(round(np.real(Nroots)))

File: Matlab/findnumberofroots.m
function Nroots = findnumberofroots(f,Z,varargin)
    
%% check for symbolic derivative of the function
    if any(strcmp(varargin,'df'))
        f1 = varargin{find(strcmp(varargin,'df'))+1};    
        option_der = 'symbolic';     % symbolic derivative has been given
    else
        option_der = 'numerical';    % second choice for derivative of function
    end
    
%% check for which way to calculate derivative
    if any(strcmp(varargin,'option_der'))
        option_der = varargin{find(strcmp(varargin,'option_der'))+1};    % if option is given then use that
    elseif strcmp(option_der,'symbolic')
        % do nothing as symbolic has been chosen either by supplying df or
        % choosing it above
    else    
        option_der = 'numerical';    % second choice for derivative of function
    end
    
%% calculate number of roots
    if strcmp(option_der,'symbolic') 
        Nroots = trapz(Z,f1(Z)./f(Z)) / (2*pi*1i);

    elseif strcmp(option_der,'numerical')
        % or numerically calculating the derivative
        dfdz = diff(f(Z))./diff(Z);

        Nroots = trapz(Z(1:end-1),dfdz./f(Z(1:end-1))) / (2*pi*1i);    % and taking 1 point less because of the function diff 
    elseif strcmp(option_der,'argp')
        % Follow derivation Peter & Kravanja for derivativeless solution
        Npoints = varargin{find(strcmp(varargin,'argp'))+3};
        Args = zeros(Npoints,1)
        for ii = 1:Npoints-1
            Args(ii) = angle(f(Z(ii+1)/f(Z(ii))));
        end
        Args(Npoints) = angle(f(Z(1)/f(Z(Npoints))));
        
        Nroots = 1/2/pi*sum(Args);

    elseif strcmp(option_der,'rootsofunity')
        z0 = varargin{find(strcmp(varargin,'rootsofunity'))+1};
        r0 = varargin{find(strcmp(varargin,'rootsofunity'))+2};
        Npoints = varargin{find(strcmp(varargin,'rootsofunity'))+3};
        
        % determine Npoints based on a step size, aka circumference divided
        % by the step size. however keep it to a maximum value 
        step_size = 1e-6;
        Npoints = round(2*pi*r0/step_size);
        if Npoints > 1e4
            Npoints = 1e4;
        end

        k = 0:1:Npoints-1;
        Z = z0 + r0*exp(2*pi*1i.*k/Npoints);
        n = Npoints;
        fk = f(Z);
        c = fft(fk)/n;
        cp = (1:n-1).*c(2:end);
        ppzk = n*ifft([cp 0])/r0;
        
        % by Austin / Kravanja
        Nroots = (real(mean(Z.*ppzk./fk)));
        % Delves and lynes but new paper approach of derivative
%         Nroots = trapz(Z,ppzk./fk) / (2*pi*1i);
        % fully numerical..
%         Nroots = trapz(Z(1:end-1),(diff(fk)./diff(Z))./fk(1:end-1)) / (2*pi*1i);
    end
%%

%     % round to nearest integer
%     Nroots = round(Nroots);
    
    % check for NaN or Inf values
    if isnan(Nroots) || isinf(Nroots)
        return
    end
    
    % IF WE CROSS THE BRANCH CUT ONCE WE GET AN IMAGINARY VALUE OF 1/PI
    % IF WE CROSS THE BRANCH CUT TWICE WE GET AN IMAGINARY VALUE OF 2/PI
    % ETC...

    % check for imaginary value because that means we crossed a
    % branch cut
    if round(imag(Nroots)/(1/pi)) ~= 0
        % check how often 1/pi fits in the imaginary part and round it to
        % get an integer which is the amount of times we cross a branch cut
        N_BC = round(imag(Nroots)/(1/pi));
        
        % if the rounded real part is also equal to zero then we know there
        % aren't any roots (or branch points) so we just stop the search
        if round(real(Nroots)) == 0
            Nroots = 0;
            return
        end
        

%         if mod(real(Nroots),0)-mod(real(Nroots),1) ~= 0 && round(mod(real(Nroots),1)) == 0
%             % when we have whole integer values return also N_BC so we know
%             % we must evade the branch cut (or use it later for something
%             % else)
%             Nroots = round(real(Nroots),2,'significant') + 1i*N_BC;
        
        % check if we have integer values or not and then round to the correct
        % values (i.e. either 0.5, 1.5, 2.5, etc.) and return that value 
        if mod(real(Nroots),0)-mod(real(Nroots),1) ~= 0
            % when we have values of 1.5, 2.5, etc
            Nroots = round(real(Nroots),2,'significant') + 1i*N_BC ;
        else
            % when we have a values of 0.5 (thus one branch point)
            Nroots = round(real(Nroots),1,'significant') + 1i*N_BC;
        end

        return
    end

    % check whether the rounded value goes to zero, if so, make it zero and
    % return that value
    if round(Nroots) == 0
        Nroots = 0;
        return
    end

    % if the above is not true, round to nearest integer
    Nroots = round(real(Nroots));
end